File: src/led/ColorRGBA.java
package led;

public class ColorRGBA {
	int r = 255;
	int g = 255;
	int b = 255;
	int a = 255;
	
	public ColorRGBA(int _r, int _g, int _b, int _a) {
		r = _r;
		g = _g;
		b = _b;
		a = _a;
	}
	
	public ColorRGB toRGB() {
		return new ColorRGB(
				(255 - a) * ColorRGB.black.r + a * r,
		        (255 - a) * ColorRGB.black.g + a * g,
		        (255 - a) * ColorRGB.black.b + a * b);
	}
	
	public ColorRGB toRGB(ColorRGB _baseColor) {
		ColorRGB rgb = new ColorRGB(
				(int)((1 - a/255.0f) * _baseColor.r + (a/255.0f) * r),
				(int)((1 - a/255.0f) * _baseColor.g + (a/255.0f) * g),
				(int)((1 - a/255.0f) * _baseColor.b + (a/255.0f) * b));
		rgb.r = Math.max(0, Math.min(rgb.r, 255));
		rgb.g = Math.max(0, Math.min(rgb.g, 255));
		rgb.b = Math.max(0, Math.min(rgb.b, 255));
		return rgb;
	}
	
	public ColorRGBA dim(float _percentage) {
		return new ColorRGBA(
				(int)(r * _percentage), 
				(int)(g * _percentage), 
				(int)(b * _percentage),
				(int)(a * _percentage));
	}
	
	public ColorRGBA dim(float _percentageR, float _percentageG, float _percentageB, float _percentageA) {
		return new ColorRGBA(
				(int)(r * _percentageR), 
				(int)(g * _percentageG), 
				(int)(b * _percentageB),
				(int)(a * _percentageA));
	}
	
	@Override
	public String toString() {
		return String.format("R%d G%d B%d A%d", r, g, b, a);
	}
}

File: src/led/ColorRGBA.kt
package led

import led.ColorRGB
import com.google.gson.annotations.SerializedName
import led.ColorChannel
import led.ColorRGBA
import led.ColorHSV
import interpolation.InterpolationType
import led.LEDStripManager
import procedures.ProcedureCalls
import javax.swing.JLabel
import led.LEDStripConfig
import led.LEDStrip
import procedures.ProcContainer
import led.LEDStripRenderer
import java.lang.InterruptedException
import procedures.Procedure
import led.json.LEDStatus
import network.LEDChangeModel
import led.LEDDataBundle
import procedures.ProcedureType
import procedures.JavascriptProcedure
import procedures.ProcedureFactory

class ColorRGBA(_r: Int, _g: Int, _b: Int, _a: Int) {
    var r = 255
    var g = 255
    var b = 255
    var a = 255
    fun toRGB(): ColorRGB {
        return ColorRGB(
            (255 - a) * ColorRGB.Companion.black.r + a * r,
            (255 - a) * ColorRGB.Companion.black.g + a * g,
            (255 - a) * ColorRGB.Companion.black.b + a * b
        )
    }

    fun toRGB(_baseColor: ColorRGB): ColorRGB {
        val rgb = ColorRGB(
            ((1 - a / 255.0f) * _baseColor.r + a / 255.0f * r).toInt(),
            ((1 - a / 255.0f) * _baseColor.g + a / 255.0f * g).toInt(),
            ((1 - a / 255.0f) * _baseColor.b + a / 255.0f * b).toInt()
        )
        rgb.r = Math.max(0, Math.min(rgb.r, 255))
        rgb.g = Math.max(0, Math.min(rgb.g, 255))
        rgb.b = Math.max(0, Math.min(rgb.b, 255))
        return rgb
    }

    fun dim(_percentage: Float): ColorRGBA {
        return ColorRGBA(
            (r * _percentage).toInt(),
            (g * _percentage).toInt(),
            (b * _percentage).toInt(),
            (a * _percentage).toInt()
        )
    }

    fun dim(_percentageR: Float, _percentageG: Float, _percentageB: Float, _percentageA: Float): ColorRGBA {
        return ColorRGBA(
            (r * _percentageR).toInt(),
            (g * _percentageG).toInt(),
            (b * _percentageB).toInt(),
            (a * _percentageA).toInt()
        )
    }

    override fun toString(): String {
        return String.format("R%d G%d B%d A%d", r, g, b, a)
    }

    init {
        r = _r
        g = _g
        b = _b
        a = _a
    }
}